tests/org.eclipse.xtext.tests/src/org/eclipse/xtext/generator/trace/TraceRegionToStringTest.xtend

```diff
@@ -0,0 +1,210 @@
+		val tester = new TraceRegionToStringTester
+		tester.local = '''
+			foo bar
+		'''
+		tester.localFrame = new TextRegion(1, 5)
+		tester.remote1 = '''
+			baz buz
+		'''
+		tester.trace = new Region(0, 7, #[new Location(1, 5, tester.uri1)]) => [
+			addChild(2, 1, #[new Location(1, 5)])
+			addChild(5, 1, #[new Location(1, 5)])
+		]
+		tester === '''
+			---- local1 ----- | ---- remote1 ----
+			o[1[o]1] b[2[a]2] | b[1,2[az bu]2,1]z
+			-------------------------------------
+			1: D 2-1 Region -> Location[1,5]
+			2: D 5-1 Region -> Location[1,5]
+		'''
+	}
+	
+	@Test def void twoFrame2() {
+		val tester = new TraceRegionToStringTester
+		tester.local = '''
+			foo bar
+		'''
+		tester.localFrame = new TextRegion(2, 4)
+		tester.remote1 = '''
+			baz buz
+		'''
+		tester.trace = new Region(0, 7, #[new Location(1, 5, tester.uri1)]) => [
+			addChild(2, 1, #[new Location(1, 5)])
+			addChild(5, 1, #[new Location(1, 5)])
+		]
+		tester === '''
+			---- local1 ---- | ---- remote1 ----
+			[1[o]1] b[2[a]2] | b[1,2[az bu]2,1]z
+			------------------------------------
+			1: D 2-1 Region -> Location[1,5]
+			2: D 5-1 Region -> Location[1,5]
+		'''
+	}
+
+	@Test def void twoFiles() {
+		val tester = new TraceRegionToStringTester
+		tester.local = '''
+			foo bar
+		'''
+		tester.remote1 = '''
+			baz buz
+		'''
+		tester.remote2 = '''
+			xxx zzz
+		'''
+		tester.trace = new Region(0, 7, #[new Location(0, 7, tester.uri1)]) => [
+			addChild(0, 3, #[new Location(0, 3)])
+			addChild(4, 3, #[new Location(0, 3, tester.uri2)])
+		]
+		tester === '''
+			------- local1 -------- | ---- remote1 -----
+			[1,2[foo]2] [3[bar]3,1] | [1,2[baz]2] buz]1]
+			                        | ---- remote2 -----
+			                        | [3[xxx]3] zzz
+			--------------------------------------------
+			1: D 0-7 Region -> Location[0,7,remote1] {
+			2: D 0-3   Region -> Location[0,3]
+			3: D 4-3   Region -> Location[0,3,remote2]
+			1:       }
+		'''
+	}
+
+	def void ===(TraceRegionToStringTester actual, CharSequence expectation) {
+		val rendered = actual.toString
+		Assert.assertEquals(expectation.toString.trim, rendered)
+	}
+
+}
+
+@Accessors class TraceRegionToStringTester extends TraceRegionToString {
+	val URI uri1 = URI.createURI("remote1")
+	val URI uri2 = URI.createURI("remote2")
+	String local
+	String remote1
+	String remote2
+	ITextRegion localFrame
+
+	new() {
+		localURI = URI.createURI("local1")
+		showLegend = false
+	}
+
+	override protected getFileInfo(URI uri) {
+		switch uri {
+			case localURI: new TracedFileInfo => [text = local.trim; frame = localFrame]
+			case uri1: new TracedFileInfo => [text = remote1.trim]
+			case uri2: new TracedFileInfo => [text = remote2.trim]
+		}
+	}
+
+}
+
+class Location extends LocationData {
+
+	new(int offset, int length, URI path) {
+		super(offset, length, 0, 0, path)
+	}
+
+	new(int offset, int length) {
+		this(offset, length, null)
+	}
+
+}
+
+class Region extends TraceRegion {
```

Comment:
Could we somehow avoid to introduce new toplevel types with heavily overloaded simple names?